### PR TITLE
OneDrive mount with elFinder

### PR DESCRIPTION
**Pull request by rajasharma27**
mount onedrive with elfinder

### CODE REVIEW

**Comment by rajasharma27** [comment]
connector configuration
---------------------------

elFinder::$netDrivers['onedrive'] = 'OneDrive';
 //include_once dirname(__FILE__).DIRECTORY_SEPARATOR.'elFinderVolumeOneDrive.class.php';
// // AND reuire regist redirect url to "YOUR_CONNECTOR_URL"

 define('ELFINDER_ONEDRIVE_CLIENTID',     'xxxxxxxxxxxxxxxxxxxxxx');
 define('ELFINDER_ONEDRIVE_CLIENTSECRET', 'xxxxxxxxxxxxxxxxxxx');

 if (substr_count($_GET['code'],'-') == 4 && $_GET['protocol'] !== 'onedrive') { 	
	$callback = new elFinderVolumeOneDrive();	
	$callback->netmountPrepare(array('onedrive_callback'=>'true'));  //prepare callback url	
 	exit();	
  }

**Comment by rajasharma27** [comment]
autoload.php
---------------------

'elFinderVolumeOneDrive' => ELFINDER_PHP_ROOT_PATH . '/elFinderVolumeOneDrive.class.php',

**Comment by rajasharma27** [comment]
elfinder .js  configuration
-----------------

onedrive: {
				name : 'OneDrive',
				inputs: {
					offline  : $('<input type="checkbox"/>').on('change', function() {
						$(this).parents('table.elfinder-netmount-tb').find('select:first').trigger('change', 'reset');
					}),
					host     : $('<span><span class="elfinder-info-spinner"/></span><input type="hidden"/>'),
					path     : $('<input type="text" value="root"/>'),
					user     : $('<input type="hidden"/>'),
					pass     : $('<input type="hidden"/>')
				},
				select: function(fm, ev, data){
					var f = this.inputs, oline = f.offline,
						data = data || null;
					if ($(f.host[0]).find('span.elfinder-info-spinner').length || data === 'reset') {
						if (oline.parent().children().length === 1) {
							f.path.parent().prev().html(fm.i18n('folderId'));
							oline.attr('title', fm.i18n('offlineAccess'));
							oline.uniqueId().after($('<label/>').attr('for', oline.attr('id')).html(' '+fm.i18n('offlineAccess')));
						}
						$(f.host[0]).empty().addClass('elfinder-info-spinner')
							.parent().find('span.elfinder-button-icon').remove();
						fm.request({
							data : {cmd : 'netmount', protocol: 'onedrive', host: 'onedrive.com', user: 'init', options: {id: fm.id, offline: oline.prop('checked')? 1:0, pass: f.host[1].value}},
							preventDefault : true
						}).done(function(data){
							$(f.host[0]).removeClass("elfinder-info-spinner").html(data.body.replace(/\{msg:([^}]+)\}/g, function(whole,s1){return fm.i18n(s1,'OneDrive.com');}));
						}).fail(function(){});
					} else {
						oline.parent().parent()[f.user.val()? 'hide':'show']();
					}
				},
				done: function(fm, data){
					var f = this.inputs, p = this.protocol;
					if (data.mode == 'makebtn') {
						$(f.host[0]).removeClass('elfinder-info-spinner');
						f.host.find('input').hover(function(){$(this).toggleClass('ui-state-hover');});
						$(f.host[1]).val('');
						f.path.val('root').next().remove();
						f.user.val('');
						f.pass.val('');
						f.offline.parent().parent().show();
					} else {
						$(f.host[0]).html('OneDrive.com&nbsp;').removeClass('elfinder-info-spinner');
						if (data.reset) {
							p.trigger('change', 'reset');
							return;
						}
						$(f.host[0]).parent().append($('<span class="elfinder-button-icon elfinder-button-icon-reload" title="'+fm.i18n('reAuth')+'">')
							.on('click', function() {
								$(f.host[1]).val('reauth');
								p.trigger('change', 'reset');
							}));
						$(f.host[1]).val('onedrive');
						if (data.folders) {
							f.path.next().remove().end().after(
								$('<div/>').append(
									$('<select class="ui-corner-all" style="max-width:200px;">').append(
										$($.map(data.folders, function(n,i){return '<option value="'+i+'">'+fm.escape(n)+'</option>'}).join(''))
									).on('change', function(){f.path.val($(this).val());})
								)
							);
						}
						f.user.val('done');
						f.pass.val('done');
						f.offline.parent().parent().hide();
					}
				},
				fail: function(fm, err){
					this.protocol.trigger('change', 'reset');
				}
			},

**Comment by rajasharma27** [comment]
elfinder .css configuration
-----------------------------

.elfinder-navbar-root-onedrive .elfinder-navbar-icon { background-image:url("../img/volume_icon_onedrive.png"); }

icon link - http://www.onlinewebplace.com/controlpanel/filemanager/files/Images/volume_icon_onedrive.png

File: php/elFinderVolumeOneDrive.class.php [hunk 1]
```diff
@@ -0,0 +1,1797 @@
...
+    protected function parseRaw($raw)
+    {		
+        $stat = array();
+	
+		$stat['rev']	= $raw->id;
+        $stat['name']	= $raw->name;
+        $stat['mime']	= $raw->type =='folder' ? 'directory' : parent::$mimetypes[pathinfo($raw->name, PATHINFO_EXTENSION)];		
+        $stat['size']	= $raw->type =='folder' ? 0 : (int)$raw->size;
+        $stat['ts']		= $raw->updated_time !== null ? strtotime($raw->updated_time) : $_SERVER['REQUEST_TIME'];
+        $stat['dirs']	= $raw->type =='folder' ? 1 : 0;
```
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: Please correctly implement the function `_subdir()`.

File: php/elFinderVolumeOneDrive.class.php [hunk 2]
```diff
@@ -0,0 +1,1797 @@
...
+    * Create thumnbnail and return it's URL on success
+    *
+    * @param  string  $path  file path
+    * @param  string  $mime  file mime type
+
+    * @return string|false
+    * @author Dmitry (dio) Levashov
+    * @author Naoki Sawada
+    **/
+    protected function createTmb($path, $stat)
```
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: Is this necessary? In enough elFinderVolumeDriver::createTmb()?

File: php/elFinderVolumeOneDrive.class.php [hunk 3]
```diff
@@ -0,0 +1,1797 @@
...
+				$res = $this->query($parentId);
+				foreach($res as $f){		
+					if($f->name == $name){					
+						return $path.'/'.$f->id();
+						break;
+					}
+				}
+			}
+            //Create or Update a file            				
+			$content = stream_get_contents($fp);
```
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: Can be handled by the stream?

File: php/elFinderVolumeOneDrive.class.php [hunk 4]
```diff
@@ -0,0 +1,1797 @@
...
+    {			
+        $opts = array(
+            'client_id'         => '',
+            'client_secret'     => '',
+            'accessToken'       => '',            
+            'root'              => 'OneDrive.com',
+            'OneDriveApiClient'   => '',
+            'path'              => '/',
+            'separator'         => '/',
+            'tmbPath'           => '../files/.tmb',
```
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: Please remove 'tmbPath', 'tmbURL' and 'tmpPath'.

File: php/elFinderVolumeOneDrive.class.php [hunk 5]
```diff
@@ -0,0 +1,1797 @@
...
+        
+        $this->options['root'] == '' ?  $this->options['root']= 'OneDrive.com' : $this->options['root'];
+        
+        if (empty($this->options['alias'])) {
+		    $this->options['alias'] = ($this->options['path'] === '/') || ($this->options['path'] === 'root')? $this->options['root'] : $this->getPathtoName($this->options['path']).'@OneDrive';
+        }
+
+        $this->rootName = $this->options['alias'];
+               
+        $this->tmbPrefix = 'onedrive'.base_convert($this->root, 10, 32);
```
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: There is a problem if you mounted multiple accounts.

File: php/elFinderVolumeOneDrive.class.php [hunk 6]
```diff
@@ -0,0 +1,1797 @@
...
+            CURLOPT_AUTOREFERER    => true,
+            CURLOPT_POST           => 1,
+            CURLOPT_POSTFIELDS     => $fields,
+
+            CURLOPT_HTTPHEADER => array(
+                'Content-Length: ' . strlen($fields),
+            ),
+
+            // SSL options.
+            CURLOPT_SSL_VERIFYHOST => false,
```
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: This is not a good idea. It must set to `2`.

File: php/elFinderVolumeOneDrive.class.php [hunk 7]
```diff
@@ -0,0 +1,1797 @@
...
+            CURLOPT_SSL_VERIFYPEER => false,
```
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: This is not a good idea. It must set to `true`.

**Comment by rajasharma27** [comment]
not getting
There is a problem if you mounted multiple accounts.

File: php/elFinderVolumeOneDrive.class.php (hunk 5 above)
Comment (nao-pon) [inline php/elFinderVolumeOneDrive.class.php]: The file name of the thumbnail seems to overlap when you mount the same name of the path. Also, it seems both thumbnails are deleted when you unmount one of the volume. But I haven't been test. I'm just reading your code.

Comment (rajasharma27) [inline php/elFinderVolumeOneDrive.class.php]: after checked your demo page mount with google drive,  tmb  create only one thumbnail for same name volume

File: php/elFinderVolumeOneDrive.class.php (hunk 2 above)
Comment (rajasharma27) [inline php/elFinderVolumeOneDrive.class.php]: not getting this point

**Comment by rajasharma27** [comment]
sir, please add with elfinder

**Pull request merged**